compiler/noirc_evaluator/src/ssa/function_builder/mod.rs

```diff
@@ -259,9 +260,118 @@ impl FunctionBuilder {
         arguments: Vec<ValueId>,
         result_types: Vec<Type>,
     ) -> Cow<[ValueId]> {
+        if let Value::Intrinsic(intrinsic) = &self.current_function.dfg[func] {
+            if intrinsic == &Intrinsic::WrappingShiftLeft {
+                let result_type = self.current_function.dfg.type_of_value(arguments[0]);
+                let bit_size = match result_type {
+                    Type::Numeric(NumericType::Signed { bit_size })
+                    | Type::Numeric(NumericType::Unsigned { bit_size }) => bit_size,
+                    _ => {
+                        unreachable!("ICE: Truncation attempted on non-integer");
+                    }
+                };
+                return self
+                    .insert_wrapping_shift_left(arguments[0], arguments[1], bit_size)
+                    .results();
+            }
+        }
+
         self.insert_instruction(Instruction::Call { func, arguments }, Some(result_types)).results()
     }
 
+    /// Insert ssa instructions which computes lhs << rhs by doing lhs*2^rhs
+    pub(crate) fn insert_shift_left(&mut self, lhs: ValueId, rhs: ValueId) -> ValueId {
+        let base = self.field_constant(FieldElement::from(2_u128));
+        let pow = self.pow(base, rhs);
+        let typ = self.current_function.dfg.type_of_value(lhs);
+        let pow = self.insert_cast(pow, typ);
+        self.insert_binary(lhs, BinaryOp::Mul, pow)
+    }
+
+    /// Insert ssa instructions which computes lhs << rhs by doing lhs*2^rhs
+    /// and truncate the result to bit_size
+    fn insert_wrapping_shift_left(
+        &mut self,
+        lhs: ValueId,
+        rhs: ValueId,
+        bit_size: u32,
+    ) -> InsertInstructionResult {
```

Comment:
What was the advantage of having this as a builtin again?

From the comment I'd expect it to output an equivalent of the old `crate::from_field(crate::as_field(x) * 2.pow_32(crate::as_field(y)))`. There's a check if the rhs is constant below, though I'd expect the previous version's pow to also be optimized if y is constant. I remember the previous method was increasing compile-times, but I'm not sure where it was from specifically. Was the lack of an early constant rhs check really the full cause of the performance regression?

Reply:
Ssa instruction count goes from: 44189634 (with the stdlib old implementation) down to 7732820 (with the builtin)
Circuit size is the same.
compilation time is much better with the builtin.
I am not sure of the root cause for the improvement, it could be because of all the casting between the types.
I tried the builtin without the constant folding and I get the same result, which makes sense because computing the power with a constant value is also optimised to a constant value.

Reply:
I'd still like to know why we're getting this large instruction count difference. Could it be not the shift left function but the pow function in the stdlib that is the issue? Perhaps we'd want a builtin pow instead in that case.

Reply:
I think the advantage lies inside the shift-left rather than the pow.
We check for the overflow before doing the computation, so we can do the pow as field without any additional check for overflow or any truncate to the type. We don't event care about field overflow, whatever the pow computes (allowing any overflow), if we shift too many bits, we simply returns 0.
We cannot make a builtin pow that computes a wrong result because it does not care about overflows. However we can do this within the shift left.